File: erlang/functions/list_operations.erl
-module(list_operations).
-export([demo/0]).

%% List operations in Erlang
%% Demonstrating functional programming with lists

demo() ->
    io:format("~n=== List Operations in Erlang ===~n~n"),

    Numbers = [1, 2, 3, 4, 5],
    Fruits = ["apple", "banana", "cherry"],

    io:format("Numbers: ~p~n", [Numbers]),
    io:format("Fruits: ~p~n", [Fruits]),

    %% List length
    io:format("~nLength: ~p~n", [length(Numbers)]),

    %% Head (first element)
    [Head | _] = Numbers,
    io:format("First element (head): ~p~n", [Head]),

    %% Tail (rest of list)
    [_ | Tail] = Numbers,
    io:format("Rest of list (tail): ~p~n", [Tail]),

    %% Prepend element
    io:format("~nPrepend 0: ~p~n", [[0 | Numbers]]),

    %% Append lists
    io:format("Append [6,7,8]: ~p~n", [Numbers ++ [6, 7, 8]]),

    %% Reverse list
    io:format("Reversed: ~p~n", [lists:reverse(Numbers)]),

    %% Member check
    io:format("~nMember 3: ~p~n", [lists:member(3, Numbers)]),
    io:format("Member 10: ~p~n", [lists:member(10, Numbers)]),

    %% Map function over list
    io:format("~nMap (square): ~p~n", [lists:map(fun(X) -> X * X end, Numbers)]),

    %% Filter list
    io:format("Filter (even): ~p~n",
              [lists:filter(fun(X) -> X rem 2 =:= 0 end, Numbers)]),

    %% Fold (reduce) left
    io:format("~nFold left (sum): ~p~n",
              [lists:foldl(fun(X, Acc) -> X + Acc end, 0, Numbers)]),

    io:format("Fold left (product): ~p~n",
              [lists:foldl(fun(X, Acc) -> X * Acc end, 1, Numbers)]),

    %% Fold right
    io:format("Fold right (list): ~p~n",
              [lists:foldr(fun(X, Acc) -> [X | Acc] end, [], Numbers)]),

    %% Find element
    case lists:search(fun(X) -> X > 3 end, Numbers) of
        {value, Val} -> io:format("~nFirst > 3: ~p~n", [Val]);
        false -> io:format("~nNo element > 3~n")
    end,

    %% Any and all
    io:format("Any > 4: ~p~n", [lists:any(fun(X) -> X > 4 end, Numbers)]),
    io:format("All positive: ~p~n", [lists:all(fun(X) -> X > 0 end, Numbers)]),

    %% Sort list
    Unsorted = [5, 2, 8, 1, 9, 3],
    io:format("~nUnsorted: ~p~n", [Unsorted]),
    io:format("Sorted: ~p~n", [lists:sort(Unsorted)]),

    %% Nth element (1-indexed in Erlang)
    io:format("~nThird element (nth): ~p~n", [lists:nth(3, Numbers)]),

    %% Take first n elements
    io:format("First 3 (sublist): ~p~n", [lists:sublist(Numbers, 3)]),

    %% Drop first n elements
    io:format("Drop 2 (nthtail): ~p~n", [lists:nthtail(2, Numbers)]),

    %% Split list
    {Left, Right} = lists:split(3, Numbers),
    io:format("~nSplit at 3: {~p, ~p}~n", [Left, Right]),

    %% Remove duplicates
    WithDupes = [1, 2, 2, 3, 3, 3, 4],
    io:format("~nWith duplicates: ~p~n", [WithDupes]),
    io:format("Unique (usort): ~p~n", [lists:usort(WithDupes)]),

    %% Zip two lists
    Letters = [a, b, c, d, e],
    io:format("~nZip: ~p~n", [lists:zip(Numbers, Letters)]),

    %% Unzip
    Zipped = [{1, a}, {2, b}, {3, c}],
    {Nums, Lets} = lists:unzip(Zipped),
    io:format("Unzip: {~p, ~p}~n", [Nums, Lets]),

    %% Flatten nested lists
    Nested = [[1, 2], [3, 4], [5]],
    io:format("~nFlatten: ~p~n", [lists:flatten(Nested)]),

    %% Partition
    {Evens, Odds} = lists:partition(fun(X) -> X rem 2 =:= 0 end, Numbers),
    io:format("~nPartition even:~n"),
    io:format("  Evens: ~p~n", [Evens]),
    io:format("  Odds: ~p~n", [Odds]),

    %% List comprehension
    Squares = [X * X || X <- [1, 2, 3, 4, 5]],
    io:format("~nList comprehension (squares): ~p~n", [Squares]),

    EvenSquares = [X * X || X <- [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], X rem 2 =:= 0],
    io:format("Even squares: ~p~n", [EvenSquares]),

    %% Pattern matching in list comprehension
    People = [{person, "Alice", 30}, {person, "Bob", 25}, {person, "Charlie", 35}],
    Names = [Name || {person, Name, _Age} <- People],
    io:format("~nNames from people: ~p~n", [Names]),

    %% Helper functions demo
    io:format("~n=== Helper Functions ===~n"),
    io:format("Sum: ~p~n", [sum(Numbers)]),
    io:format("Product: ~p~n", [product(Numbers)]),
    io:format("Maximum: ~p~n", [maximum(Numbers)]),
    io:format("Minimum: ~p~n", [minimum(Numbers)]),
    io:format("Range 1-10: ~p~n", [range(1, 10)]),
    io:format("Take 3: ~p~n", [take(3, Numbers)]),
    io:format("Drop 2: ~p~n", [drop(2, Numbers)]),

    ok.

%% Helper functions

sum(List) ->
    lists:foldl(fun(X, Acc) -> X + Acc end, 0, List).

product(List) ->
    lists:foldl(fun(X, Acc) -> X * Acc end, 1, List).

maximum([H | T]) ->
    lists:foldl(fun erlang:max/2, H, T).

minimum([H | T]) ->
    lists:foldl(fun erlang:min/2, H, T).

range(Start, End) when Start >= End ->
    [];
range(Start, End) ->
    [Start | range(Start + 1, End)].

take(0, _) ->
    [];
take(_, []) ->
    [];
take(N, [H | T]) when N > 0 ->
    [H | take(N - 1, T)].

drop(0, List) ->
    List;
drop(_, []) ->
    [];
drop(N, [_ | T]) when N > 0 ->
    drop(N - 1, T).

File: fsharp/functions/list_operations.fs
// List operations in F#
// Demonstrating functional programming with immutable lists

module ListOperations

open System

// Demo function
let demo () =
    printfn "\n=== List Operations in F# ===\n"

    // Creating lists
    let numbers = [1; 2; 3; 4; 5]
    let fruits = ["apple"; "banana"; "cherry"]

    printfn "Numbers: %A" numbers
    printfn "Fruits: %A" fruits

    // List length
    printfn "\nLength: %d" (List.length numbers)

    // Head (first element)
    printfn "First element (head): %d" (List.head numbers)

    // Tail (rest of list)
    printfn "Rest of list (tail): %A" (List.tail numbers)

    // Prepend element (cons operator)
    printfn "\nPrepend 0: %A" (0 :: numbers)

    // Append lists
    printfn "Append [6;7;8]: %A" (numbers @ [6; 7; 8])

    // Reverse list
    printfn "Reversed: %A" (List.rev numbers)

    // Contains check
    printfn "\nContains 3: %b" (List.contains 3 numbers)
    printfn "Contains 10: %b" (List.contains 10 numbers)

    // Find element
    let found = List.tryFind (fun x -> x > 3) numbers
    printfn "\nFirst > 3: %A" found

    // Map function over list
    printfn "\nMap (square): %A" (List.map (fun x -> x * x) numbers)

    // Filter list
    printfn "Filter (even): %A" (List.filter (fun x -> x % 2 = 0) numbers)

    // Fold (reduce) left
    printfn "\nFold left (sum): %d" (List.fold (+) 0 numbers)
    printfn "Fold left (product): %d" (List.fold (*) 1 numbers)

    // Fold right
    printfn "Fold right (list): %A" (List.foldBack (fun x acc -> x :: acc) numbers [])

    // Exists and forall
    printfn "\nExists (> 4): %b" (List.exists (fun x -> x > 4) numbers)
    printfn "For all (> 0): %b" (List.forall (fun x -> x > 0) numbers)

    // Sort list
    let unsorted = [5; 2; 8; 1; 9; 3]
    printfn "\nUnsorted: %A" unsorted
    printfn "Sorted: %A" (List.sort unsorted)
    printfn "Sorted descending: %A" (List.sortDescending unsorted)

    // Take and skip
    printfn "\nTake 3: %A" (List.take 3 numbers)
    printfn "Skip 2: %A" (List.skip 2 numbers)

    // Zip two lists
    let letters = ['a'; 'b'; 'c'; 'd'; 'e']
    printfn "\nZip: %A" (List.zip numbers letters)

    // Unzip
    let zipped = [(1, 'a'); (2, 'b'); (3, 'c')]
    let (nums, lets) = List.unzip zipped
    printfn "Unzip: (%A, %A)" nums lets

    // Partition
    let (evens, odds) = List.partition (fun x -> x % 2 = 0) numbers
    printfn "\nPartition even:"
    printfn "  Evens: %A" evens
    printfn "  Odds: %A" odds

    // Group by
    let grouped = List.groupBy (fun x -> x % 3) [1..9]
    printfn "\nGroup by (mod 3): %A" grouped

    // Remove duplicates
    let withDupes = [1; 2; 2; 3; 3; 3; 4]
    printfn "\nWith duplicates: %A" withDupes
    printfn "Distinct: %A" (List.distinct withDupes)

    // Chunk
    printfn "\nChunk 3: %A" (List.chunkBySize 3 [1..10])

    // Windowed
    printfn "Windowed 3: %A" (List.windowed 3 numbers)

    // List comprehension
    let squares = [ for x in 1..5 -> x * x ]
    printfn "\nList comprehension (squares): %A" squares

    let evenSquares = [ for x in 1..10 do
                          if x % 2 = 0 then
                              yield x * x ]
    printfn "Even squares: %A" evenSquares

    // Range
    printfn "\nRange [1..10]: %A" [1..10]
    printfn "Range [1..2..10]: %A" [1..2..10]  // step by 2

    // Pattern matching on lists
    let rec listSum lst =
        match lst with
        | [] -> 0
        | head :: tail -> head + listSum tail

    printfn "\nList sum (pattern matching): %d" (listSum numbers)

    // Helper functions
    printfn "\n=== Helper Functions ==="
    printfn "Sum: %d" (sum numbers)
    printfn "Product: %d" (product numbers)
    printfn "Maximum: %d" (maximum numbers)
    printfn "Minimum: %d" (minimum numbers)
    printfn "Average: %.2f" (average numbers)

// Helper functions

let sum lst = List.fold (+) 0 lst

let product lst = List.fold (*) 1 lst

let maximum lst = List.max lst

let minimum lst = List.min lst

let average lst =
    let total = List.sum lst |> float
    total / float (List.length lst)

let range start stop =
    [start..stop]

let take n lst = List.take n lst

let drop n lst = List.skip n lst

let flatten lst = List.collect id lst

let unique lst = List.distinct lst

// Run demo
[<EntryPoint>]
let main argv =
    demo()
    0
